cpp/models/abm/analyze_result.h

```diff
@@ -56,8 +56,15 @@ std::vector<Model> ensemble_params_percentile(const std::vector<std::vector<Mode
             single_element[run] = get_param(params);
         }
         std::sort(single_element.begin(), single_element.end());
-        auto& new_params = get_param(percentile[n]);
-        new_params       = single_element[static_cast<size_t>(num_runs * p)];
+        if constexpr (std::is_same_v<decltype(get_param(percentile[n])), double>) {
+            // If get_param returns a double, just use the value
+            get_param(percentile[n]);
+        }
```

Comment:
Can you explain what this part is doing? `get_param(percentile[n])` returns a reference to the parameters of the output model that contains all percentile parameters. Thus, I can't see what this is doing.

Reply:
In the code, we function `param_percentil` which takes a function `get_param` as a parameter and uses it to calculate a percentile value. The `get_param` function is supposed to return a reference to a parameter of a model, which is then used to calculate and assign a new value.

The issue arises in this line:

`auto& new_params = get_param(percentile[n]);`

This line expects `get_param` to return a reference to a double (or another type), which can be modified. However, in some cases, `get_param` might return a temporary value, which cannot be bound to a non-const lvalue reference, leading to a compilation error. Examples are ViralLoadDistributions, GoToWorkTimeMaximum, ... 

For example, the `days()` function in `get<LockdownDate>().days()` returns a temporary double value, which cannot be bound to a non-const lvalue reference. Similar to ViralLoadDistributions, InfectivityDistribution.... 

Using `if constexpr`, you can create a compile-time check to handle different cases based on the return type of `get_param`. If it returns a double, it compiles and runs the first block, treating `new_params` as a temporary value.
If it returns a reference, it compiles and runs the second block, allowing us to modify the original parameter.

Reply:
For me, `auto& testp =
        infection_params
            .get<mio::abm::ViralLoadDistributions>()[{mio::abm::VirusVariant::Wildtype, mio::abm::AgeGroup::Age0to4}];` for example works.
You say that the `new_params` is used as a temporary value in the first block, but there is just one command `get_param(percentile[n]);`that doesn't do any assignment, does it? That's what I meant with my comment that I don't see what it does. :)

For the "special" parameters such as ViralLoadDistributions that don't have a comparator that can be used for sorting, I would suggest that we don't use this function anyway, since we need to access a deeper structure to compare. To be honest, I would have expected that these parameters fail the sorting in the first place.

Reply:
Yes, I realised also that it did not do anything. As a temporary solution for the code to compile, I added a static cast to the result of these parameter calls. Maybe I could also add a future issue for this as well.